### PR TITLE
Helm gcs gateway support

### PR DESCRIPTION
**Pull request by dhughesni**
## Description
- These changes have been migrated from the previous chart: https://github.com/helm/charts/tree/master/stable/minio
- Added `gcs` support for gateway mode.
- Added new gcs block under the gateway key to house the gcs specific variables.
- The gateway-deployment template now sets the env var: GOOGLE_APPLICATION_CREDENTIALS as a path to the gcs_key.json
- The gcs_key.json can be added to the minio secret if an existingSecret is not specified.

## Motivation and Context
- The current chart does not enable gateway mode to be set as `gcs`. We need the ability to deploy minio as a gcs gateway via helm.
- Minio Slack Reference: https://minio.slack.com/archives/C3NDUB8UA/p1637607661036000

## How to test this PR?
```
$ cd helm/minio/
```
### modify values for gcs gateway deployment
```
$ git diff
diff --git a/helm/minio/values.yaml b/helm/minio/values.yaml
index 604beb39e..50c2ac171 100644
--- a/helm/minio/values.yaml
+++ b/helm/minio/values.yaml
@@ -32,7 +32,7 @@ mcImage:
 ## Distributed MinIO ref: https://docs.minio.io/docs/distributed-minio-quickstart-guide
 ## NAS Gateway MinIO ref: https://docs.min.io/docs/minio-gateway-for-nas.html
 ## GCS Gateway MinIO ref: https://docs.min.io/docs/minio-gateway-for-gcs.html
-mode: distributed
+mode: gateway
 
 ## NOTE: currently only "nas,gcs" gateways are supported.
 
@@ -106,11 +106,11 @@ pools: 1
 
 # Deploy if 'mode == gateway' - 4 replicas.
 gateway:
-  type: "nas" # currently only "nas,gcs" are supported.
-  replicas: 4
+  type: "gcs" # currently only "nas,gcs" are supported.
+  replicas: 1
   gcs:
-    gcsKeyJson: "" # credential json file of service account key (not required if using existingSecret)
-    projectId: "" # Google cloud project id required
+    gcsKeyJson: "<GCS_JSON_SERVICE_ACCOUNT>" # credential json file of service account key
+    projectId: "<GCS_PROJECT_ID>" # Google cloud project id
 
 ## TLS Settings for MinIO
 tls:
@@ -130,7 +130,7 @@ trustedCertsSecret: ""
 ## ref: http://kubernetes.io/docs/user-guide/persistent-volumes/
 ##
 persistence:
-  enabled: true
+  enabled: false
```

### debug install as dry-run to check entrypoint is configured correctly
```
$ helm install miniogcs . --debug --dry-run | grep "minio gateway gcs"
...
(standard input):496:            - "/usr/bin/docker-entrypoint.sh minio gateway gcs dr-exploration-team -S /etc/minio/certs/ --address :9000 --console-address :9001 "
```

### test chart install
```
$ helm install minio .
NAME: minio
LAST DEPLOYED: Thu Dec  2 15:57:07 2021
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
Minio can be accessed via port 9000 on the following DNS name from within your cluster:
minio.default.svc.cluster.local
To access Minio from localhost, run the below commands:
  1. export POD_NAME=$(kubectl get pods --namespace default -l "release=minio" -o jsonpath="{.items[0].metadata.name}")
  2. kubectl port-forward $POD_NAME 9000 --namespace default
Read more about port forwarding here: http://kubernetes.io/docs/user-guide/kubectl/kubectl_port-forward/
You can now access Minio server on http://localhost:9000. Follow the below steps to connect to Minio server with mc client:
  1. Download the Minio mc client - https://docs.minio.io/docs/minio-client-quickstart-guide
  2. export ACCESS_KEY=$(kubectl get secret minio -o jsonpath="{.data.rootUser}" | base64 --decode) 
  3. export SECRET_KEY=$(kubectl get secret minio -o jsonpath="{.data.rootPassword}" | base64 --decode)
  4. mc alias set minio-local http://localhost:9000 "$ACCESS_KEY" "$SECRET_KEY" --api s3v4
  5. mc ls minio-local
Alternately, you can use your browser or the Minio SDK to access the server - https://docs.minio.io/categories/17

$ export POD_NAME=$(kubectl get pods --namespace default -l "release=minio" -o jsonpath="{.items[0].metadata.name}")
$ kubectl port-forward $POD_NAME 9000 --namespace default &
[1] 26269
Forwarding from 127.0.0.1:9000 -> 9000
Forwarding from [::1]:9000 -> 9000
$ export ACCESS_KEY=$(kubectl get secret minio -o jsonpath="{.data.rootUser}" | base64 --decode) 
$ export SECRET_KEY=$(kubectl get secret minio -o jsonpath="{.data.rootPassword}" | base64 --decode)
$ mc alias set minio-local http://localhost:9000 "$ACCESS_KEY" "$SECRET_KEY" --api s3v4
mc: Configuration written to `/Users/daryl.hughes/.mc/config.json`. Please update your access credentials.
mc: Successfully created `/Users/daryl.hughes/.mc/share`.
mc: Initialized share uploads `/Users/daryl.hughes/.mc/share/uploads.json` file.
mc: Initialized share downloads `/Users/daryl.hughes/.mc/share/downloads.json` file.
Added `minio-local` successfully.
$ mc ls minio-local
Handling connection for 9000
... Testing validated: GCS object store buckets are listed confirming minio is configured as gcs gateway
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [X] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
Is this still draft @dhughesni ?

**Comment by dhughesni** [comment]
> 

@harshavardhana  Yes still a draft atm.
I plan to do some final testing before the end of this week - at that point I will convert to PR  ready for review 👍

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by dhughesni** [comment]
@harshavardhana PR is ready for review - please let me know if there are any further actions required from me.

File: helm/minio/templates/gateway-deployment.yaml [hunk 1]
```diff
@@ -95,6 +105,10 @@ spec:
                 secretKeyRef:
                   name: {{ template "minio.secretName" . }}
                   key: rootPassword
+            {{- if eq .Values.gateway.type "gcs" }}
+            - name: GOOGLE_APPLICATION_CREDENTIALS
+              value: "/etc/credentials/gcs_key.json"
```
Comment (harshavardhana) [inline helm/minio/templates/gateway-deployment.yaml]:
```suggestion
              value: "/tmp/gcs-credentials/service-account-file.json""
```

File: helm/minio/templates/gateway-deployment.yaml [hunk 2]
```diff
@@ -69,8 +69,18 @@ spec:
...
+            {{- if eq .Values.gateway.type "nas" }}
             - "/usr/bin/docker-entrypoint.sh minio gateway nas {{ $bucketRoot }} -S {{ .Values.certsPath }} --address :{{ .Values.minioAPIPort }} --console-address :{{ .Values.minioConsolePort }} {{- template "minio.extraArgs" . }} "
+            {{- end }}
+            {{- if eq .Values.gateway.type "gcs" }}
+            - "/usr/bin/docker-entrypoint.sh minio gateway gcs {{ .Values.gateway.gcs.projectId }} -S {{ .Values.certsPath }} --address :{{ .Values.minioAPIPort }} --console-address :{{ .Values.minioConsolePort }} {{- template "minio.extraArgs" . }} "
+            {{- end }}
           volumeMounts:
+            {{- if eq .Values.gateway.type "gcs" }}
+            - name: minio-user
+              mountPath: "/etc/credentials"
```
Comment (harshavardhana) [inline helm/minio/templates/gateway-deployment.yaml]:
```suggestion
              mountPath: "/tmp/gcs-credentials"
```

File: helm/minio/templates/secrets.yaml [hunk 3]
```diff
@@ -13,4 +13,7 @@ type: Opaque
 data:
   rootUser: {{ if .Values.rootUser }}{{ .Values.rootUser | toString | b64enc | quote }}{{ else }}{{ randAlphaNum 20 | b64enc | quote }}{{ end }}
   rootPassword: {{ if .Values.rootPassword }}{{ .Values.rootPassword | toString | b64enc | quote }}{{ else }}{{ randAlphaNum 40 | b64enc | quote }}{{ end }}
+  {{- if eq .Values.gateway.type "gcs" }}
+  gcs_key.json: {{ .Values.gateway.gcs.gcsKeyJson | b64enc }}
```
Comment (harshavardhana) [inline helm/minio/templates/secrets.yaml]:
```suggestion
  service-account-file.json: {{ .Values.gateway.gcs.serviceAccountFile | b64enc }}
```

File: helm/minio/values.yaml [hunk 4]
```diff
@@ -105,8 +106,11 @@ pools: 1
 
 # Deploy if 'mode == gateway' - 4 replicas.
 gateway:
-  type: "nas" # currently only "nas" is supported.
+  type: "nas" # currently only "nas,gcs" are supported.
   replicas: 4
+  gcs:
+    gcsKeyJson: "" # credential json file of service account key (not required if using existingSecret)
```
Comment (harshavardhana) [inline helm/minio/values.yaml]:
```suggestion
    serviceAccountFile: "" # credential JSON file of service account key (not required if using existing secret)
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :x: [more...](#af2426d6-4eba-4c38-a950-53e78bbd73fe) |
<a name="af2426d6-4eba-4c38-a950-53e78bbd73fe"></a>
### 13810-c32f7e3/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s4:32314
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 1 minutes and 56 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 28 seconds
(8/15) Running minio-dotnet tests ... done in 56 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 23 seconds
(10/15) Running minio-java tests ... done in 29 seconds
(11/15) Running minio-js tests ... FAILED in 54 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook for \"removeObject(bucketName, objectName, removeOpts)",
  "args": "bucketName:minio-js-test-ver-presignef3cbc4d-e5aa-4a1f-86d9-36853dcd1aae ",
  "duration": 5,
  "alert": "datafile-1-b ",
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty. You must delete all versions in the bucket. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:89:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:164:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:670:10) at endReadableNT (internal/streams/readable.js:1334:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(11/15) Running minio-py tests ... done in 1 minutes and 13 seconds
(12/15) Running s3cmd tests ... done in 25 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Pull request merged**